lib/user_session_management.rb

```diff
@@ -105,4 +108,43 @@ def signup_email
     signup_state.try(:contact_info_value)
   end
 
+  def set_client_app(client_id)
+    @client_app = client_id.nil? ?
+                    nil :
+                    Doorkeeper::Application.find_by(uid: client_id)
+    session[:client_app] = @client_app.present? ? @client_app.id : nil
+  end
+
+  def get_client_app
+    @client_app ||= session[:client_app].nil? ?
+                      nil :
+                      Doorkeeper::Application.find_by(id: session[:client_app])
+  end
+
+  def set_alternate_signup_url(url)
+    if url.blank?
+      session[:alt_signup] = nil
+    elsif !url.is_a?(String)
+      raise IllegalArgument
+    elsif is_redirect_url?(application: get_client_app, url: url)
+      session[:alt_signup] = url
+    else
+      session[:alt_signup] = nil
```

Comment:
Could move these ifs into `is_redirect_url?` but meh...

Would simplify the logic in this function though...